Add distributed workload test tutorial (#5629)
opensearch-project/documentation-website: _benchmark/user-guide/distributed-load.md

---
layout: default

> title: Configuring OpenSearch Benchmark

Author: ```suggestion
title: Running distributed loads
```

nav_order: 10
parent: User guide
---

> # Running distributed load generation 

Author: ```suggestion
# Running distributed loads 
```

> By default, OpenSearch Benchmark runs loads on the same machine where the benchmark started. However, you can generate addiional benchmarking load, particularly on large clusters with multiple machines, you can use multiple load drivers. This page details how to distribute benchmark loads across multiple machines in a single cluster.

Author: ```suggestion
By default, OpenSearch Benchmark runs loads on the same machine where the benchmark started. However, you can generate additional benchmarking testing loads, particularly on large clusters with multiple machines, you can use multiple load drivers. This page details how to distribute benchmark loads across multiple machines in a single cluster.
```

## System architecture 

The following tutorial uses a three-node architecture; each node is generated in [Amazon Elastic Compute Cloud (Amazon EC2)](https://docs.aws.amazon.com/ec2/?nc2=h_ql_doc_ec2):

- **Node 1**: Node 1 acts as the _coordinator node_ and enables distribution and communication between the other two nodes.
- **Node 2** and **Node 3**: The remaining nodes in the cluster are used to generate the load for the benchmark test.

OpenSearch Benchmark must be installed on all nodes. For installation instructions, see [Installing OpenSearch Benchmark]({{site.url}}{{site.baseurl}}/benchmark/user-guide/installing-benchmark/).

Make note of each node's IP address. This tutorial uses the following IP addresses:

> **Node 1 -- Coordinator node**: 192.0.1.0

Reviewer: Should these be formatted as a bulleted list?

**Node 2 -- Worker node**: 198.52.100.0
**Node 3 -- Worker node**: 198.53.100.0

## Step 1: Enable node communication

Make sure to enable communication for each node. In the AWS Management Console:

1. Go to the EC2 host for the node.
2. Select **Security**, and then select the security group associated with the node. 
3. Use **Add inbound rules** to open traffic to the node, based on the port range and traffic type of your cluster.

## Step 2: Run daemon processes on each node

> Start OpenSearch Benchmark on each node, using the `--node-ip` to initialize OpenSearch Benchmark on the node itself and then `--coordinator-ip` to connect each node to the coordinator node.

Reviewer: ```suggestion
Start OpenSearch Benchmark on each node, using `--node-ip` to initialize OpenSearch Benchmark on the node itself and then `--coordinator-ip` to connect each node to the coordinator node.
```

> Now, start OpenSearch Benchmark on each node, using the `--node-ip` to intailize OpenSearch Benchmark on the node itself, then `--coordinator-ip` to connect each node to the Coordinator node.

Author: ```suggestion
Now, start OpenSearch Benchmark on each node, using the `--node-ip` to initialize OpenSearch Benchmark on the node itself, then `--coordinator-ip` to connect each node to the Coordinator node.
```

> For **Node 1**, the following command identifies the node as coordinator:

Reviewer 2: ```suggestion
For **Node 1**, the following command identifies the node as coordinator:
```
_the_ coordinator?

```

> opensearch-benchmarkd start —node-ip=192.0.1.0 —coordinator-ip=192.0.1.0

Reviewer 3: Need to ensure that the parameters use `--` for the beginning. For example, `—node-ip` should be `--node-ip` and `—coordinator-ip` should be `--coordinator-ip`.

> opensearch-benchmarkd start —node-ip=192.0.1.0 —coordinator-ip=192.0.1.0

Author: ```suggestion
opensearch-benchmarkd start --node-ip=192.0.1.0 --coordinator-ip=192.0.1.0
```

```

> The following commands enable **Node 2** and **Node 3** to listen to the coordinator node for load generation instructions:

Reviewer: ```suggestion
The following commands enable **Node 2** and **Node 3** to listen to the coordinator node for load generation instructions.
```

**Node 2**

```

> opensearch-benchmarkd start —node-ip=198.52.100.0 —coordinator-ip=192.0.1.0

Reviewer 3: `—node-ip` --> `--node-ip` 
`—coordinator-ip` --> `--coordinator-ip`

> opensearch-benchmarkd start —node-ip=198.52.100.0 —coordinator-ip=192.0.1.0

Author: ```suggestion
opensearch-benchmarkd start --node-ip=198.52.100.0 --coordinator-ip=192.0.1.0
```

```

**Node 3**

```

> opensearch-benchmarkd start —node-ip=198.53.100.0 —coordinator-ip=192.0.1.0

Reviewer 3: `—node-ip` --> `--node-ip` 
`—coordinator-ip` --> `--coordinator-ip`

> opensearch-benchmarkd start —node-ip=198.53.100.0 —coordinator-ip=192.0.1.0

Author: ```suggestion
opensearch-benchmarkd start --node-ip=198.53.100.0 --coordinator-ip=192.0.1.0
```

```

> With OpenSearch Benchmark running on all three nodes and the worker nodes set to listen to the coordinator node, you can now run the Benchmark test.

Author: ```suggestion
With OpenSearch Benchmark running on all three nodes and the worker nodes set to listen to the coordinator node, you can now run the benchmark test.
```

> With OpenSearch Benchmark running on all three nodes and the worker nodes set to listen to the coordinator node, you can now run the Benchmark test.

Reviewer 2: Move this statement below the heading, "Run the benchmark test"?

> ## Step 3: Running the benchmark test

Reviewer 2: ```suggestion
## Step 3: Run the benchmark test
```

> On **Node 1**, run a benchmark test with the `load-worker-coordinator-hosts` set to the IP addresses for your worker nodes, as shown in the following example:

Reviewer 3: `load-worker-coordinator-hosts` --> `worker-ips`

These have been updated in a recent PR. Sorry for the confusion!

> On **Node 1**, run a benchmark test with the `load-worker-coordinator-hosts` set to the IP addresses for your worker nodes, as shown in the following example:

Author: ```suggestion
On **Node 1**, run a benchmark test with the `worker-ips` set to the IP addresses for your worker nodes, as shown in the following example:
```

```

> opensearch-benchmark execute_test —pipeline=benchmark-only —workload=eventdata —load-worker-coordinator-hosts=198.52.100.0,198.53.100.0 —target-hosts=<DOMAIN_ENDPOINT> —client-options=<STANDARD_CLIENT_OPTIONS> —kill-running-processes

Reviewer 3: Can we revise this to ensure that the markdown shows `--` instead of a single line? Also `—load-worker-coordinator-hosts` --> `--worker-ips`

> opensearch-benchmark execute_test —pipeline=benchmark-only —workload=eventdata —load-worker-coordinator-hosts=198.52.100.0,198.53.100.0 —target-hosts=<DOMAIN_ENDPOINT> —client-options=<STANDARD_CLIENT_OPTIONS> —kill-running-processes

Author: ```suggestion
opensearch-benchmark execute_test --pipeline=benchmark-only --workload=eventdata --worker-ips=198.52.100.0,198.53.100.0 --target-hosts=<DOMAIN_ENDPOINT> --client-options=<STANDARD_CLIENT_OPTIONS> --kill-running-processes
```

```

> After the test runs, the logs generated by the test appear on your worker nodes.

Author: ```suggestion
After the test completes, the logs generated by the test appear on your worker nodes.
```

> 

Author: I'll ask Ian about this.

> 

Author: I'll ask Ian about this.

> 

Author: No troubleshooting steps at the moment. We'll add additional ones as we go along.
